### PR TITLE
Speed up pfield computation, and fix near-field artifact

### LINKED ISSUE #192: Reduce recomputing `pfield` when `set_transmits` is used.
Currently, whenever you change the `selected_transmits` in the `Scan` class, the `pfield` has to be recomputed. We could cache the `pfield` per transmit and only do the normalization again when needed.

### PR DESCRIPTION
**Pull request by wesselvannierop**
I wanted to fix #192 but this seemed to be quite a challenge with the current infrastructure. So I ended up speeding up the pfield computation itself through parallelization. This was previously not possible because it used varying frequency bins based on every transmit. Now it uses the same amount for all transmits, based on the worst case. For high-resolution pressure fields, the speed-up can be over 60x!

You can specify the number of grid point to run in parallel through `point_batch_size`, which by defaults takes a conservative `2048`, such that it works on basically all GPUs

Furthermore, this PR fixes a bug in the implementation that previously occurred in the first row of pixels, see regression test.

Additionally, I cleaned up variable naming inside the pfield func.

**How to test**

You can use the scripts posted in the collapsed sections.

## Benchmarks

_Including compile times. Using the L40S with max batch sizes._

### Default downsample & downmix

| Version | Backend | Focused | Plane-wave | Shape |
|---|---|--:|--:|---|
| new | tensorflow | 3.605 s | 0.114 s | (128, 975, 617) |
| new | jax | 7.132 s | 2.440 s | (128, 975, 617) |
| old | tensorflow | 8.940 s | 4.679 s | (128, 975, 617) |
| old | jax | 16.531 s | 6.546 s | (128, 975, 617) |

### `downmix=1` & `downsample=1`

| Version | Backend | Focused | Plane-wave | Shape |
|---|---|--:|--:|---|
| new | tensorflow | 4.973 s | 1.244 s | (128, 975, 617) |
| new | jax | 6.829 s | 2.664 s | (128, 975, 617) |
| old | tensorflow | 75.191 s | 64.375 s | (128, 975, 617) |
| old | jax | 84.522 s | 55.289 s | (128, 975, 617) |

<details closed>
<summary>Benchmark script</summary>

```python

"""Minimal benchmark for `compute_pfield`.

Computes the pressure field once for 128 focused transmits and once for
128 plane-wave transmits and prints the wall-clock time.

Run with e.g.:
    KERAS_BACKEND=jax python benchmark_pfield.py
    KERAS_BACKEND=tensorflow python benchmark_pfield.py
"""

import os
import time

# Disable the on-disk cache so every call actually recomputes.
os.environ["ZEA_DISABLE_CACHE"] = "1"

import zea

zea.init_device()

import keras  # noqa: E402
import numpy as np  # noqa: E402

from zea.beamform.delays import (  # noqa: E402
    compute_t0_delays_focused,
    compute_t0_delays_planewave,
)
from zea.beamform.pfield import compute_pfield  # noqa: E402
from zea.probes import Verasonics_l11_4v  # noqa: E402

N_TX = 128
SOUND_SPEED = 1540.0
BANDWIDTH_PERCENT = 70.0


def make_grid(probe):
    """Build a dummy (grid_size_z, grid_size_x, 3) imaging grid."""
    wavelength = SOUND_SPEED / probe.center_frequency
    dx = dz = wavelength / 4
    x = np.arange(-19e-3, 19e-3, dx)
    z = np.arange(0.0, 60e-3, dz)
    grid_z, grid_x = np.meshgrid(z, x, indexing="ij")
    grid = np.stack([grid_x, np.zeros_like(grid_x), grid_z], axis=-1)
    return grid.astype("float32")


def make_focused_delays(probe):
    """t0_delays of shape (N_TX, n_el) for 128 focused line transmits."""
    aperture = probe.probe_geometry[:, 0]
    origins_x = np.linspace(aperture.min(), aperture.max(), N_TX)
    transmit_origins = np.stack([origins_x, np.zeros(N_TX), np.zeros(N_TX)], axis=-1)
    focus_distances = np.full(N_TX, 30e-3)
    polar_angles = np.zeros(N_TX)
    return compute_t0_delays_focused(
        transmit_origins=transmit_origins,
        focus_distances=focus_distances,
        probe_geometry=probe.probe_geometry,
        polar_angles=polar_angles,
        sound_speed=SOUND_SPEED,
    )


def make_planewave_delays(probe):
    """t0_delays of shape (N_TX, n_el) for 128 plane-wave transmits."""
    polar_angles = np.linspace(-15, 15, N_TX) * np.pi / 180
    return compute_t0_delays_planewave(
        probe_geometry=probe.probe_geometry,
        polar_angles=polar_angles,
        sound_speed=SOUND_SPEED,
    )


def benchmark(name, t0_delays, probe, grid, tx_apodizations):
    """Run compute_pfield once and report the wall-clock time."""
    start = time.perf_counter()
    result = compute_pfield(
        sound_speed=SOUND_SPEED,
        center_frequency=probe.center_frequency,
        bandwidth_percent=BANDWIDTH_PERCENT,
        n_el=probe.n_el,
        probe_geometry=probe.probe_geometry,
        tx_apodizations=tx_apodizations,
        grid=grid,
        t0_delays=t0_delays,
        downmix=1,
        downsample=1,
    )
    # Force the (possibly lazy) computation to finish before stopping the timer.
    result = np.asarray(keras.ops.convert_to_numpy(result))
    elapsed = time.perf_counter() - start
    print(f"  {name:>10s} | {elapsed:7.3f} s | shape={result.shape}")
    return elapsed


def main():
    print(f"Backend: {keras.backend.backend()}")
    probe = Verasonics_l11_4v()
    grid = make_grid(probe)
    tx_apodizations = np.ones((N_TX, probe.n_el)) * np.hanning(probe.n_el)[None]

    cases = {
        "focused": make_focused_delays(probe),
        "planewave": make_planewave_delays(probe),
    }

    print(f"N_TX={N_TX}, n_el={probe.n_el}, grid={grid.shape[:2]}")
    for case_name, t0_delays in cases.items():
        print(f"\n[{case_name}]")
        benchmark(case_name, t0_delays, probe, grid, tx_apodizations)


if __name__ == "__main__":
    main()
```
</details>

## Regression test

Comparing to the implementation on main. Using `norm=True` which makes the individual pressure fields look a bit weird, but this is to make sure every pixel gets weighted to 1.

image

<details closed>
<summary>Regression script</summary>

```python
"""Regression check: current `compute_pfield` vs the version on `main`.

Computes the pressure field for a set of plane-wave transmits with both the
working-tree implementation and the implementation currently on the `main`
branch, then writes a side-by-side plot for visual inspection.

Run with e.g.:
    KERAS_BACKEND=jax python regression_pfield.py
"""

import importlib.util
import os
import subprocess
import tempfile

# Disable the on-disk cache so both implementations actually recompute.
os.environ["ZEA_DISABLE_CACHE"] = "1"

import zea  # noqa: E402

zea.init_device()

import matplotlib.pyplot as plt  # noqa: E402
import numpy as np  # noqa: E402

from zea.beamform.delays import compute_t0_delays_planewave  # noqa: E402
from zea.beamform.pfield import compute_pfield as compute_pfield_current  # noqa: E402
from zea.probes import Verasonics_l11_4v  # noqa: E402

SOUND_SPEED = 1540.0
BANDWIDTH_PERCENT = 70.0
N_TX = 8


def load_main_compute_pfield():
    """Import `compute_pfield` from commit 9947954 as a standalone module."""
    src = subprocess.check_output(["git", "show", "9947954:zea/beamform/pfield.py"], text=True)
    tmp = tempfile.NamedTemporaryFile("w", suffix=".py", delete=False)
    tmp.write(src)
    tmp.close()
    spec = importlib.util.spec_from_file_location("pfield_main", tmp.name)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module.compute_pfield


def main(norm=True):
    compute_pfield_main = load_main_compute_pfield()

    probe = Verasonics_l11_4v()
    n_el = probe.n_el

    tx_apodizations = np.ones((N_TX, n_el)) * np.hanning(n_el)[None]
    polar_angles = np.linspace(-15, 15, N_TX) * np.pi / 180
    t0_delays = compute_t0_delays_planewave(
        probe_geometry=probe.probe_geometry,
        polar_angles=polar_angles,
        sound_speed=SOUND_SPEED,
    )

    # Modest dummy grid.
    x = np.linspace(-19e-3, 19e-3, 300)
    z = np.linspace(0.0, 55e-3, 420)
    grid_z, grid_x = np.meshgrid(z, x, indexing="ij")
    grid = np.stack([grid_x, np.zeros_like(grid_x), grid_z], axis=-1).astype("float32")

    kwargs = dict(
        sound_speed=SOUND_SPEED,
        center_frequency=probe.center_frequency,
        bandwidth_percent=BANDWIDTH_PERCENT,
        n_el=n_el,
        probe_geometry=probe.probe_geometry,
        tx_apodizations=tx_apodizations,
        grid=grid,
        t0_delays=t0_delays,
        norm=norm,
    )

    pf_current = np.asarray(compute_pfield_current(**kwargs))
    pf_main = np.asarray(compute_pfield_main(**kwargs, verbose=False))

    abs_diff = np.abs(pf_current - pf_main)
    rel = abs_diff.max() / (pf_main.max() + 1e-12)
    print(f"current shape: {pf_current.shape}, main shape: {pf_main.shape}")
    print(f"max abs diff: {abs_diff.max():.3e}  (relative to main max: {rel:.3e})")
    print(f"mean abs diff: {abs_diff.mean():.3e}")

    # Plot a few transmits: rows = [main, current, abs diff], cols = transmits.
    tx_to_plot = [0, N_TX // 3, 2 * N_TX // 3, N_TX - 1]
    extent = [grid_x.min() * 1e3, grid_x.max() * 1e3, grid_z.max() * 1e3, grid_z.min() * 1e3]

    fig, axes = plt.subplots(
        3, len(tx_to_plot), figsize=(4 * len(tx_to_plot), 10), constrained_layout=True
    )
    for col, tx in enumerate(tx_to_plot):
        for row, (label, field) in enumerate([("main", pf_main[tx]), ("current", pf_current[tx])]):
            ax = axes[row, col]
            vmax = np.percentile(field.ravel(), 99.9)
            im = ax.imshow(field, extent=extent, cmap="viridis", vmin=0, vmax=vmax)
            ax.set_title(f"{label} | tx {tx}")
            fig.colorbar(im, ax=ax, fraction=0.046)
        ax = axes[2, col]
        im = ax.imshow(abs_diff[tx], extent=extent, cmap="magma")
        ax.set_title(f"abs diff | tx {tx}")
        fig.colorbar(im, ax=ax, fraction=0.046)

    for ax in axes.ravel():
        ax.set_xlabel("x [mm]")
        ax.set_ylabel("z [mm]")

    out = "pfield_regression.png"
    fig.savefig(out, dpi=120)
    print(f"saved plot to {os.path.abspath(out)}")


if __name__ == "__main__":
    main()
```
</details>

## Release Notes

* **Refactor**
  * Optimized pressure field computation for improved performance using batched tensor processing.
  * Enhanced 3D scan operation error handling with explicit error messages.

* **Bug Fixes**
  * Improved default handling and error reporting for scan timing parameters.

* **Chores**
  * Removed verbose logging functionality.
  * Removed unused imports.

[Review Change Stack](https://app.coderabbit.ai/change-stack/tue-bmd/zea/pull/267?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by tristan-deep** [comment]
Is #350 also addressed here?

**Comment by wesselvannierop** [comment]
> Is #350 also addressed here?

Currently testing if this makes sense.

**Comment by wesselvannierop** [comment]
TODO: decrease the `point_batch_size` to `1024` to avoid OOM with `pixels_per_wavelength=4` on a 10GB GPU.

**Comment by Louisvh** [comment]
Nice - this one is much more readable than the original version.

Not a change in this PR, but I do wonder: currently the code assumes a linear array with z=0 and x=[-w/2,w/2], but it reads the pitch as the distance between the coordinates of the first two elements in probe_geometry. Doesn't it make more sense to use those coordinates to set the true z-offset from the center and calculate the element_theta values? As far as I can tell, that's all that's needed to get correct pfield estimates for curved probes, which I don't think is the case currently (but I might be missing something).

Edit: made a new issue for this - #365

**Comment by wesselvannierop** [comment]
Decided to tackle both these comments in another PR.

**Review by Louisvh — APPROVED** [review]
